### DISCUSSION TITLE
problem with mini batching the adjacency matrix created from the nodes and edge_indices

### DISCUSSION THREAD

**Discussion by akpas001** [pyg-team/pytorch_geometric#4713] [Q&A] [answered]
Hi,

I am trying to implement the concept of VGAE (variational graph autoencoder ) on a custom dataset. so, for that i have to create an adjacency matrix. The way i created my adjacency matrix is as follows.

I have 9 nodes per graph, and dynamic edge_indices. since I have 9 nodes. I have created a 9*9 matrix and checked for the active edges present.
```python
dataset = []
adj = torch.zeros(9,9)
edges = edge_index[:,:int(active_edges)] # this gives the number of active edges present in that certain graph.
for i in edges.t():
     adj[i[0]i[1]] = 1 # replacing the zeros with ones whereever an active edge is present.
     dataset.append(Data(x =x, edge_index = edge_index[:,:int(active_edges)], adj _= adj)) # appending this inro a dataloader.

train_loader = DataLoader(train_dataset, batch_size = 32)

for data in train_loader:
    x, edge_index, adj = data.x, data.edge_index, data.adj
    print(x.shape) = [288,30] # because (batch_size*num_nodes, num_features)
    print(edge_index.shape) = [2, 441] # because (coo format) and  441 can vary for different because it is dynamic.
    print(adj.shape) = [288,9] # because again (batch_size*num_nodes, 9)
```
so, now the problem is during training. when i predict my adjacency matrix, using VGAE my latent feature output shape is (288, 20)
and i pass it through Innerproduct decoder to calculate adjacency and it returns me matrix of shape of (288,288) bacause torch.matmul(lat_fea, lat_fea.t()).

Now i cannot backpropogate these two calculate my loss. I dont understand if I am missing something here. Please help.

**Comment by Padarn** [comment]
Sorry I couldn't follow your example, what do you mean
> Now i cannot backpropogate these two calculate my loss. I dont understand if I am missing something here. Please help.

Could you provide a working code block with your issue?

**Reply by akpas001** [reply]
sure. I did upload my code in the comment below. please feel free to go through it.

**Comment by rusty1s** [comment, marked as answer]
If you want to reconstruct edges batch-wise, you will need to first convert your node feature matrix of shape `(288, 20)` to `(batch_size, num_nodes, 20) = (32, 9, 20)`. Then, you can re-construct edges for each adjacency matrix:
```python
recon = x @ x.transpose(1, 2)
```

**Reply by akpas001** [reply]
I want to have the shape of the adj and adj_pred to be same. How can i achieve that? There is no problem with the nodes or edge_indices. I am posting my code for your reference so that you can run it and check.

**Reply by rusty1s** [reply]
Can you give a bit more details? In the example above, `recon` will have shape of `[batch_size, 9, 9]`, which I assume is exactly what you want :)

**Reply by akpas001** [reply]
Thankyou for taking out your time for helping me. So, basically the main objective of the VGAE model, I am implementing is to predict the adjacency matrix. so, to have a `target_adjacency `matrix, I have created an `adjacency matrix` using nodes and edge_indices. Now, when I am predicting the adjacency matrix. There is a size mismatch happening between the `predicted adjacency` `(288,288)` and `target adjacency (288,9)`. (288 is because batch_size = 32 and num_nodes = 9). The reason for this is mini batching. So, my question is during the creation of adjacency matrix using nodes and edge_indices am I missing anything? I have uploaded my code and data for your reference in the below comment. Please feel free to go through it and help me.
Thank you!

**Reply by Padarn** [reply]
I think you just want to mask your target adjacency to only include the nodes in that minibatch?

**Reply by akpas001** [reply]
masking my `target adjacency` matrix doesnot convert it into the shape `(288,288)`. Does it?

**Reply by Padarn** [reply]
Sorry was replying from my phone and missed some context.

I think what @rusty1s suggests is the best option (assuming I understand correctly): Your `target_adjacency` is a stack of `32, 9 x 9` adjacency graphs. So you can do:
```
target_adjacency = target_adjacency.reshape(32, 9, 9)
```
which will match the output share of what @rusty1s suggests
```
recon = x @ x.transpose(1, 2)
```

If you really want to turn your `target_adjacency` into the (288,288) shape, you can do something like
```
target_adjacency = torch.block_diag(*[target_adjacency[i*9:(i+1)*9, :] for i in range(32)])
```
which will build a block diagonal adjacency matrix, but this will require a lot more memory

**Reply by akpas001** [reply]
Thanks for the inputs. The first concept worked.

**Comment by akpas001** [comment]
[states.zip](https://github.com/pyg-team/pytorch_geometric/files/8786771/states.zip)
[adjacency_pred_vgae .txt](https://github.com/pyg-team/pytorch_geometric/files/8812583/adjacency_pred_vgae.txt)

Please feel free to go through the code and use the data provided, if it helps you to understand the problem batter. Thanks

**Reply by Padarn** [reply]
Thanks, I couldn't load this btw because I don't have a CUDA device.. but could read the code.

**Reply by akpas001** [reply]
Hi, 
i have updated the code and my model is not learning, the loss looks very bad. I have tried everything in my knowledge. can you please help me regarding this. I have updated the code as per the suggestion as well. you can find the updated code above. please help. Thanks!
![image](https://user-images.githubusercontent.com/56883247/171365756-b50c23ce-0ef6-41b4-abfe-c53c832408c0.png)

**Reply by Padarn** [reply]
I'd suggest you create some synthetic data that should be very easy to learn so that you can test your model against it - this will help you know if it is your model, or the problem you are trying to solve, that is the problem

**Reply by akpas001** [reply]
okay. will try that. Thanks

**Reply by akpas001** [reply]
It looks like my data is wrong. Do you have any idea among these PyTorch geometric datasets which have one hot encoded features and suitable for multi label classification task?

**Reply by akpas001** [reply]
My idea is to train a similar dataset and use the encoder as a feature extractor for my policy.

**Reply by rusty1s** [reply]
You could look into `CLUSTER` and `PATTERN` datasets. I think the might be similar.

**Reply by akpas001** [reply]
Hi,
I tried checking my model on cora dataset. It is returning `RuntimeError: Function 'AddmmBackward0' returned nan values in its 1th output`.
I happened to get the same error few days ago while implementing the similar model and using the loss function as `nn.BCELoss()` but while using `nn.BCEwithLogits` it did not throw any Nans instead gave me huge loss values. I dont understand what's creating nans. Here is the code for your reference. please help. Thanks! :)

I realize that `kl_divergence` loss being used here is throwing the `nans`.

[cora_implementation.txt](https://github.com/pyg-team/pytorch_geometric/files/8854561/cora_implementation.txt)

**Reply by rusty1s** [reply]
You can use `torch.detect_anomaly` to spot the issues of `NaN`s in your code, see [here](https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly).

**Reply by akpas001** [reply]
Thanks!:) that helped.

**Reply by akpas001** [reply]
My adjacency matrix is a sparse matrix. Do you think there is any preprocessing necessary before we send the data into the network? I have looked into the Kipf example (VGAE) implementation, where they did preprocess their data before sending it to the network. I do not exactly understand what they did with their sparse graph. If preprocessing is required could you please throw some light on it? without any preprocessing, my results are as follows. I am not able exactly to reconstruct the adjacency matrix. How do I increase the accuracy of the link prediction?

![image (4)](https://user-images.githubusercontent.com/56883247/173194846-a8ba8260-19d6-476a-83ed-728f981b1aae.png)
![image (3)](https://user-images.githubusercontent.com/56883247/173194847-495214ad-cea2-4d62-8fef-1bfb5aec5dd3.png)
![image (2)](https://user-images.githubusercontent.com/56883247/173194849-6fec20e5-9322-4768-95da-ca4199149b21.png)
![image (1)](https://user-images.githubusercontent.com/56883247/173194850-c83a16bb-8d87-4737-b274-899496a3bf17.png)

**Reply by rusty1s** [reply]
This is hard to tell. From my understanding, there is no pre-processing done in the VGAE example (except for the normalization of the adjacency matrix but which is done internally in PyG). What you can try is to use a more expressive decoder, e.g., an MLP on top of concatenated source and destination node embeddings.

**Reply by akpas001** [reply]
Sorry, that i did not update the model, This is the updated model. The  above results are from this model. Thanks for being so patient and helping all along the discussion! :)
````
class VGAE(nn.Module):
    def __init__(self):
        super(VGAE,self).__init__()
        
        # self.init_weights()
        dim1 = 30
        dim2 = 45
        dim3 = 60
        dim4 = 40
        dim5 = 20
        dim6 = 15
        dim7 = 9
        
        self.mlp1 = nn.Sequential(Linear(dim1, dim2, weight_initializer=('glorot')),\
                                  nn.LeakyReLU(), Linear(dim2, dim3))
        
        self.mlp2 = nn.Sequential(Linear(dim3, dim4),nn.Dropout(0.25),nn.LeakyReLU(),Linear(dim4, dim5))
        
        self.mlp3 = nn.Sequential(Linear(dim1, dim2, weight_initializer=('glorot')),\
                                  nn.LeakyReLU(), Linear(dim2, dim3))
                                 
        self.mlp4 = nn.Sequential(Linear(dim3, dim4),nn.Dropout(0.25),nn.LeakyReLU(),Linear(dim4, dim5))
            
        self.mlp5 = nn.Sequential(Linear(dim5, dim4),nn.LeakyReLU(),Linear(dim4, dim2))
            
        self.mlp6 = nn.Sequential(Linear(dim2, dim1),nn.Dropout(0.25),nn.LeakyReLU(),Linear(dim1, dim5))
        
        self.mlp7 = nn.Sequential(Linear(dim5, dim6),nn.LeakyReLU(),Linear(dim6, dim7))
        
        self.conv1 = GINConv(self.mlp1)
        self.conv2 = GINConv(self.mlp2)
        self.conv3 = GINConv(self.mlp3)
        self.conv4 = GINConv(self.mlp4)
        self.conv5 = GINConv(self.mlp5)
        self.conv6 = GINConv(self.mlp6)
        self.conv7 = GINConv(self.mlp7)
        
    def encode(self, x, edge_index):
        
        x = x.float()
        mu = F.relu(self.conv1(x,edge_index)) 
        mu = self.conv2(mu, edge_index)
        
        log_var = F.relu(self.conv3(x,edge_index))
        log_var = self.conv4(log_var,edge_index)

        std = torch.exp(0.5*log_var) # reparameterization
        epsilon = torch.randn_like(std)
        z = mu + epsilon*std
        
        return z, mu, log_var
    
    def decode(self,z, edge_index):

        adj_pred = self.conv5(z, edge_index)
        adj_pred = self.conv6(adj_pred, edge_index)
        adj_pred = self.conv7(adj_pred, edge_index)
        
        return adj_pred
    
    def total_loss(self, adj_pred, adj, mu, log_var):
        ns_loss = nn.BCEWithLogitsLoss()
        recons_loss = ns_loss(adj_pred, adj)
        kld_loss = -0.5 * torch.mean(torch.sum(1+2*log_var - mu**2 -\
                                               log_var.exp()**2, dim = 1))
        total_loss = recons_loss + kld_loss 
        return total_loss, recons_loss, kld_loss`
```

**Reply by rusty1s** [reply]
I am not totally sure I understand your model. It looks like your `decode` function computes new node embeddings rather than an adjacency matrix. Can you clarify? How does `ns_loss` look like?

**Reply by akpas001** [reply]
so, basically my adjacency matrix is a sparse matrix, and looks like this. (without any self nodes as well)

```
[[0, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 1, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 1, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 1, 0],
 [0, 1, 1, 1, 0, 0, 0, 0, 0],
 [1, 0, 1, 0, 0, 0, 0, 0, 0],
 [1, 1, 0, 1, 0, 0, 0, 0, 0],
 [1, 1, 1, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0]]
```
so, inorder to reproduce it, i cannot use inner product decoder as it produces ones in the diagonal, so i am just passing the latent vector through more layers to reproduce the adjacency matrix. Am I doing wrong by doing that? Please, let me know.

and coming to the `ns_loss ` it looks something like this. the problem with this is it saturates after a point. as the batch size decreases the value of ns_loss keeps going up. (because number of ietrations increases.)
![image](https://user-images.githubusercontent.com/56883247/174451951-ac8fae21-8584-4f07-8c3a-04d90cac97a8.png).


just for the reference attaching the training function as well.

```
#%% data loading
total_states = torch.load('states.pt').to(device)
#%% data processing
states = total_states[:256]

x_o = states[:,:270].reshape(-1,9,30).to(device)
edge_index_o = states[:,391:431].reshape(len(x_o),2,-1).to(torch.long).to(device)
prev_job_o = states[:,270:390].reshape(-1,4,30).to(device)
cur_mach = states[:,390].reshape(len(x_o),1).to(device)
active_edges_o = states[:,431].reshape(len(x_o),1).to(device)

dataset = []

for _x,_edg,_active_edges in\
    zip(x_o,edge_index_o,active_edges_o):
        # generating the adjacency matrix using edge indices
    adj = torch.zeros(9,9).to(device)   
    edges = _edg[:,:int(_active_edges)]
    for i in edges.t():
        adj[i[0]][i[1]] = 1
    dataset.append(Data(x =_x, edge_index = _edg[:,:int(_active_edges)],\
                        adj = adj))

#%% dataloading
random.shuffle(dataset)
split = int(len(dataset)*0.9)
train_dataset = dataset[:split]
test_dataset = dataset[split:]
test_loader = DataLoader(test_dataset, batch_size = 32, drop_last = True,\
                         shuffle = True) # only test loader is loaded.


#%% training 
model = VGAE().to(device)
model_optim = optim.Adamax(model.parameters(), lr = 8e-04, weight_decay = 1e-06)
lr_scheduler = optim.lr_scheduler.StepLR(model_optim, step_size=1000, gamma=0.01)
model.train()
torch.autograd.set_detect_anomaly(True)
epochs = 5000
tot_loss = []
tot_r_loss = []
tot_kl_loss = []
link_pred_acc = []
for ep in range(epochs):
    # print('\rEpoch Number ', ep, end = '')
    train_loader = DataLoader(train_dataset, batch_size = 32, drop_last = True,\
                              shuffle = True)
    ep_loss = 0
    ep_r_loss = 0
    ep_kl_loss = 0
    count = 0
    acc = 0 
    for data in train_loader:
        count+=1
        x, edge_index,adj = data.x, data.edge_index, data.adj
        z, mu, log_var = model.encode(x, edge_index)
        adj_pred = model.decode(z, edge_index)
        loss, r_loss, kl_loss = model.total_loss(adj_pred, adj, mu, log_var)
       
        # print('loss is ', loss)
        ep_loss += loss.item()
        # print('loss is ', loss.item())
        ep_r_loss += r_loss.item()
        ep_kl_loss += kl_loss.item()
        model_optim.zero_grad()
        loss.backward()
        
        # for name, params in model.named_parameters():
        #     print('gradients', params.grad.mean())
        #     print('weights', params.data)
        model_optim.step()
# prediciting the probabaility of the link between the nodes
    rounded_pred = torch.round(torch.sigmoid(adj_pred)).detach()
    link = torch.mul(rounded_pred, adj)
    link_prob = ((link.sum())/(adj.sum())).cpu().numpy()

    link_pred_acc.append(link_prob)
    tot_loss.append(ep_loss)
    tot_r_loss.append(ep_r_loss)
    tot_kl_loss.append(ep_kl_loss)
    lr_scheduler.step()
    if ep % 100 == 0:
        print("Epoch:", '%04d' % (ep + 1), "train_loss=", "{:.5f}".format(ep_loss),\
                "recon_loss=", "{:.5f}".format(ep_r_loss),"time=", "{:.5f}".format(time.time() - t))


plt.plot(tot_loss)
plt.show()

plt.title('recons_loss')
plt.plot(tot_r_loss)
plt.show()

plt.title('kl_loss')
plt.plot(tot_kl_loss)
plt.show()

plt.title('link_prob')
plt.plot(link_pred_acc)
plt.show()
```

**Reply by rusty1s** [reply]
You should be able to use an inner product decoder, but not train against self-loop edges:
```python
adj_pred[torch.arange(N), torch.arange(N)] = 0.
```

**Reply by akpas001** [reply]
hi, i did try with that and the result is the same as before. i am attaching the model here for reference. what i think is sinc ethe matrix is sparse does it need any auxilary operations on it for the network to predict better? 

```
class VGAE(nn.Module):
    def __init__(self):
        super(VGAE,self).__init__()
        
        # self.init_weights()
        input_dim = 30
        dim2 = 45
        dim3 = 60
        dim4 = 40
        dim5 = 20
        dim6 = 15
        output_dim = 9
        
        self.mlp1 = nn.Sequential(Linear(input_dim, dim2, weight_initializer=('glorot')),\
                                  nn.LeakyReLU(), Linear(dim2, dim3))
        
        self.mlp2 = nn.Sequential(Linear(dim3, dim4),nn.Dropout(0.25),nn.LeakyReLU(),Linear(dim4, dim5))
        
        self.mlp3 = nn.Sequential(Linear(input_dim, dim2, weight_initializer=('glorot')),\
                                  nn.LeakyReLU(), Linear(dim2, dim3))
                                 
        self.mlp4 = nn.Sequential(Linear(dim3, dim4),nn.Dropout(0.25),nn.LeakyReLU(),Linear(dim4, dim5))
            
        # self.mlp5 = nn.Sequential(Linear(dim5, dim4),nn.LeakyReLU(),Linear(dim4, dim2))
            
        # self.mlp6 = nn.Sequential(Linear(dim2, input_dim),nn.Dropout(0.25),nn.LeakyReLU(),Linear(input_dim, dim5))
        
        # self.mlp7 = nn.Sequential(Linear(dim5, dim6),nn.LeakyReLU(),Linear(dim6, output_dim))
        
        self.conv1 = GINConv(self.mlp1)
        self.conv2 = GINConv(self.mlp2)
        self.conv3 = GINConv(self.mlp3)
        self.conv4 = GINConv(self.mlp4)
        # self.conv5 = GINConv(self.mlp5)
        # self.conv6 = GINConv(self.mlp6)
        # self.conv7 = GINConv(self.mlp7)
        
    def encode(self, x, edge_index):
        
        x = x.float()
        mu = F.relu(self.conv1(x,edge_index)) 
        mu = self.conv2(mu, edge_index)
        
        log_var = F.relu(self.conv3(x,edge_index))
        log_var = self.conv4(log_var,edge_index)

        std = torch.exp(0.5*log_var) # reparameterization
        epsilon = torch.randn_like(std)
        z = mu + epsilon*std
        
        return z, mu, log_var
    
    def decode(self,z):
        
        # adj_pred = self.conv5(z, edge_index)
        # adj_pred = self.conv6(adj_pred, edge_index)
        # adj_pred = self.conv7(adj_pred, edge_index)
        z = z.reshape(-1,9,20)
        adj_pred = torch.tensor([])
        for idx, ele in enumerate(z):
            ad = torch.matmul(z[idx],z[idx].t())
            ad[torch.arange(9), torch.arange(9)] = 0
            if len(adj_pred)== 0:
                adj_pred = ad
            else:
                adj_pred = torch.cat((adj_pred,ad), dim = 0)
        
        return adj_pred
    
    def total_loss(self, adj_pred, adj, mu, log_var):
        ns_loss = nn.BCEWithLogitsLoss()
        recons_loss = ns_loss(adj_pred, adj)
        kld_loss = -0.5 * torch.mean(torch.sum(1+2*log_var - mu**2 -\
                                               log_var.exp()**2, dim = 1))
        total_loss = recons_loss + kld_loss 
        return total_loss, recons_loss, kld_loss
```
![image](https://user-images.githubusercontent.com/56883247/175574550-b2984114-ecf9-4664-9486-cf4b89c7f503.png)
![image](https://user-images.githubusercontent.com/56883247/175574607-2b8210c1-0733-4911-a5f1-6c64efc41491.png)
![image](https://user-images.githubusercontent.com/56883247/175574652-2dd37ba7-6f45-4cde-9fd5-2687dd004cbb.png)
![image](https://user-images.githubusercontent.com/56883247/175574715-c9e7c953-ce51-4ca6-beb0-02bc75d89cf3.png)

**Reply by Padarn** [reply]
Have you tried simplifying your model (reducing layers) and increasing inner dimensions? Personally I find it easiest if I can at least show that I can overfit to my data, and then adding complications or information bottlenecks.

**Reply by akpas001** [reply]
All I use is 2 layers for predicting `mu` and 2 layers to predict `log_var` and I am sampling the size of hidden layers from optuna. i personally feel there is something i am missing in the `decode ` or the network is not powerful enough to recreate the sparse matrix and somee auxilary operations are required to be done to the sparse matrix in order for it to be easily predictable.

**Reply by rusty1s** [reply]
If you feel your `decode` is not powerful enough, there exists more expressive solutions, e.g., concatenating source and destination node features and inputting them into an MLP.

**Reply by akpas001** [reply]
> You should be able to use an inner product decoder, but not train against self-loop edges:
> 
> ```python
> adj_pred[torch.arange(N), torch.arange(N)] = 0.
> ```

with this i can only eliminate self loops but i still have the problem of other elements of the matrix mirroring along the diagonal as well. We still couldn't be able to solve that issue right? So, that's why i thought rather than using the inner product decoder i better use the new convolution layers to recreate it.

**Reply by rusty1s** [reply]
What do you mean with other elements along the diagonal? Any other diagonal does not relate to a self-loop. Note that you cannot use a GNN to decode as it can only learn a permutation-sensitive mapping by mapping to `num_nodes` output channels. IMO, you should input your source and destination node features into an MLP to predict the score of an edge:
```python
src_x = x[edge_index[0]]
dst_x = x[edge_index[1]]
return mlp(torch.cat([src_x, dst_x], dim=-1))
```

**Reply by akpas001** [reply]
Sorry, that I was not clear earlier. So basically, my `target adjacency matrix ` is 
```
[[0, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 1, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 1, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 1, 0],
 [0, 1, 1, 1, 0, 0, 0, 0, 0],
 [1, 0, 1, 0, 0, 0, 0, 0, 0],
 [1, 1, 0, 1, 0, 0, 0, 0, 0],
 [1, 1, 1, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0]]
```
As it is clear from this that there are not elements mirroring along the main diagonal of the matrix. So in case if i use a inner product decoder `torch.matmul(z,z.t())` will reproduce a matrix that can mirror along the main diagonal (this works in case of bi directional graph networks) but my case is uni directional. Let us assume an my encode is outputting z as follows.

```
tensor([[ 0.7402, -0.9026, -1.4435,  1.5756, -0.0658,  0.3879,  0.6596,  0.7467,
          1.0986],
        [-0.9424, -0.3512,  0.6380,  1.1355, -0.7336,  1.1429,  1.5982,  0.5788,
         -0.9317],
        [ 0.8738, -0.4022,  0.2918, -1.4139, -1.1425, -1.3502,  0.3453, -1.8113,
         -0.1754],
        [ 0.4874,  0.2017,  0.4148, -0.2863,  0.2889, -0.2628,  0.2983, -1.8473,
          1.5490],
        [-0.1697,  1.7242,  1.0939, -0.9265,  0.1451,  1.3983,  1.0892,  0.3728,
         -2.4734],
        [-0.9506,  0.5948, -1.5473, -0.9600, -1.1656, -0.0752, -0.2214,  0.5881,
          1.0726],
        [ 0.2585, -1.7752, -0.3056, -0.0352, -0.5942, -0.3016, -0.8219, -1.0371,
         -2.4113],
        [-0.7311,  1.0770,  0.5511,  0.0124, -1.3417,  1.1180, -0.2399,  0.7053,
          0.9806],
        [-1.1700,  0.5284, -0.7145,  0.0207,  0.0169,  0.6876,  1.5358, -0.4081,
          0.5332]]
```
`adj_pred = torch.matmul(z,z.t())`
```
tensor([[ 8.2827,  1.4418, -3.4050, -0.4729, -5.9082,  0.9994, -1.8640, -0.3217,
          1.2807],
        [ 1.4418,  8.3095, -3.1398, -3.1386,  4.9531, -1.6339,  0.5688,  2.0498,
          2.9797],
        [-3.4050, -3.1398,  9.5691,  4.0729, -1.1317, -0.0606,  4.0039, -2.4377,
         -1.2444],
        [-0.4729, -3.1386,  4.0729,  6.5859, -3.5366, -0.5186, -2.5056, -0.4511,
          1.0962],
        [-5.9082,  4.9531, -1.1317, -3.5366, 14.4757, -2.5652,  0.7681,  1.5173,
          1.4744],
        [ 0.9994, -1.6339, -0.0606, -0.5186, -2.5652,  7.4829, -3.0941,  3.4705,
          2.4326],
        [-1.8640,  0.5688,  4.0039, -2.5056,  0.7681, -3.0941, 11.3220, -4.7084,
         -3.3650],
        [-0.3217,  2.0498, -2.4377, -0.4511,  1.5173,  3.4705, -4.7084,  6.5649,
          1.6436],
        [ 1.2807,  2.9797, -1.2444,  1.0962,  1.4744,  2.4326, -3.3650,  1.6436,
          5.4416]]
```
`rounded_pred = torch.sigmoid(adj_pred).round()`

```
tensor([[1., 1., 0., 0., 0., 1., 0., 0., 1.],
        [1., 1., 0., 0., 1., 0., 1., 1., 1.],
        [0., 0., 1., 1., 0., 0., 1., 0., 0.],
        [0., 0., 1., 1., 0., 0., 0., 0., 1.],
        [0., 1., 0., 0., 1., 0., 1., 1., 1.],
        [1., 0., 0., 0., 0., 1., 0., 1., 1.],
        [0., 1., 1., 0., 1., 0., 1., 0., 0.],
        [0., 1., 0., 0., 1., 1., 0., 1., 1.],
        [1., 1., 0., 1., 1., 1., 0., 1., 1.]]

```
removing self loops will produce 

```
tensor([[0., 1., 0., 0., 0., 1., 0., 0., 1.],
        [1., 0., 0., 0., 1., 0., 1., 1., 1.],
        [0., 0., 0., 1., 0., 0., 1., 0., 0.],
        [0., 0., 1., 0., 0., 0., 0., 0., 1.],
        [0., 1., 0., 0., 0., 0., 1., 1., 1.],
        [1., 0., 0., 0., 0., 0., 0., 1., 1.],
        [0., 1., 1., 0., 1., 0., 0., 0., 0.],
        [0., 1., 0., 0., 1., 1., 0., 0., 1.],
        [1., 1., 0., 1., 1., 1., 0., 1., 0.]]
```
Now, if we observe this matrix, it is already producing elements wheich are being mirrored along the main diagonal which i donot have in the `target_adjacency matrix`. so, that's why i think it might not work in my case as my case is uni directional graph. Do you know any approaches to predict such matrices ? Thanks for your patience!

**Reply by rusty1s** [reply]
If your graph is undirected, you can still do something like `self.mlp(x_src * x_dst)` (hadamard product).

**Reply by akpas001** [reply]
my graph is a directed graph, but information only process in a single direction, from source to target.

**Reply by akpas001** [reply]
I just found a paper for `Directed Graph Autoencoders ` where they are implementing `Asymmetric Inner Prooduct Decoder` for the decoder.
I am still going through it, it looks a bit complicated for me. Will get back to you if I can find some solution there. Else again will be back with my doubts. :D. I really appriciate your help!
https://www.aaai.org/AAAI22Papers/AAAI-5979.KolliasG.pdf

**Reply by akpas001** [reply]
According to this paper they are outputting two latent vectors from the encoder one for source and the other for target and calculating crossproduct between `z_s`,` z_t.(t)`. How is that possible, i am having a problem when trying to do that with my edge_index. My model looks something of this sort . It's returning me an error on edge_index saying

`RuntimeError: index 4 is out of bounds for dimension 0 with size 4`.

please Help!

```
import numpy as np
import random
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch_geometric.loader import DataLoader
from torch_geometric.data import Data
from torch_geometric.nn import Linear, Sequential, GINConv, VGAE, GCNConv
from torch_geometric.nn.models import InnerProductDecoder
from dataloader import target_adj, states
import matplotlib.pyplot as plt
import time

# device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
device = torch.device('cpu')
batch_size = 32
t = time.time()
#%%
class VGAE(nn.Module):
    def __init__(self):
        super(VGAE,self).__init__()
        
        dim1 = 30
        dim2 = 45
        dim3 = 60
        dim4 = 40
        dim5 = 20
        dim6 = 15
        dim7 = 9
        
        self.mlp1 = nn.Sequential(Linear(dim1, dim2, weight_initializer=('glorot')),\
                                  nn.ReLU(), Linear(dim2, dim3))
                                 
        self.mlp2 = nn.Sequential(Linear(dim3, dim4),nn.Tanh(),Linear(dim4, dim5))
        
        self.mlp3 = nn.Sequential(Linear(dim1, dim2, weight_initializer=('glorot')),\
                                  nn.ReLU(), Linear(dim2, dim3))
                                 
        self.mlp4 = nn.Sequential(Linear(dim3, dim4),nn.ReLU(),Linear(dim4, dim5))
            
        self.mlp5 = nn.Sequential(Linear(dim5, dim4),nn.ReLU(),Linear(dim4, dim2))
            
        self.mlp6 = nn.Sequential(Linear(dim2, dim1),nn.ReLU(),Linear(dim1, dim5))
        
        self.mlp7 = nn.Sequential(Linear(dim5, dim6),nn.ReLU(),Linear(dim6, dim7))
        
        self.conv1 = GINConv(self.mlp1)
        self.conv2 = GINConv(self.mlp2)
        self.conv3 = GINConv(self.mlp3)
        self.conv4 = GINConv(self.mlp4)
        self.conv5 = GINConv(self.mlp5)
        self.conv6 = GINConv(self.mlp6)
        self.conv7 = GINConv(self.mlp7)
    def encode(self, data):
        
        x = data[:270].reshape(9,30).float()   
        edge_index = data[391:431].reshape(2,-1).long()
        act_edges = int(data[431])
        edge_index = edge_index[:,:int(act_edges)]
        
        src_x = x[edge_index[0]]
        tar_x = x[edge_index[1]]
        x = x.float()
        mu_src = self.conv1(src_x,edge_index) 
        log_var_src = nn.ReLU()(self.conv2(mu_src, edge_index))
        mu_tar = self.conv3(x,edge_index)
        log_var_tar = nn.ReLU()(self.conv4(log_var_src,edge_index))
        std_src = torch.exp(0.5*log_var_src) # reparameterization
        std_tar = torch.exp(0.5*log_var_tar)
        epsilon = torch.randn_like(std_src)
        z_src = mu_src + epsilon*std_src
        z_tar = mu_tar + epsilon*std_tar
        return z_src, z_tar, mu_src, log_var_src, mu_tar, log_var_tar
    
    def dot_prod_decode(self,z_src, z_tar):
        
        # edge_index = data[391:431].reshape(2,-1).long()
        # act_edges = int(data[431])
        # edge_index = edge_index[:,:int(act_edges)]
        
        # adj_pred = self.conv5(z, edge_index)
        # adj_pred = self.conv6(adj_pred, edge_index)
        # adj_pred = self.conv7(adj_pred, edge_index)
        adj_pred = torch.matmul(z_src,z_tar.t())

        return adj_pred
        
    def total_loss(self, adj_pred, adj, mu, log_var ):
        ns_loss = nn.BCEWithLogitsLoss()
        kld_loss = -0.5 * torch.mean(torch.sum(1+2*log_var - mu**2 -\
                                                log_var.exp()**2, dim = 1))
        recons_loss = ns_loss(adj_pred, adj)
        # re_loss = reg_loss(mu, log_var)
        total_loss = recons_loss + kld_loss  
        
        return total_loss, recons_loss, kld_loss 

#%% training. Not being able to shuffle data, gradients might be memorized. 
model = VGAE().to(device)
model_optim = optim.Adam(model.parameters(), lr = 8e-04)
model.train()
torch.autograd.set_detect_anomaly(True)
epochs = 250
tot_loss = []
tot_r_loss = []
tot_mse_loss = []
for ep in range(epochs):
    print('\rEpoch Number ', ep, end = '')
    ep_loss = 0
    ep_recon_loss = 0
    ep_reg_loss = 0
    count = 0
    for data, adj in zip(states, target_adj):
        count+=1
        # print('\rietration number is', count, end = '')
        z_src, z_tar,mu_src, log_var_src, mu_tar, log_var_tar = model.encode(data)
        adj_pred = model.dot_prod_decode(z_src,z_tar)
        loss, recons_loss, re_loss = model.total_loss(adj_pred, adj, mu_src,\
                                                      log_var_src, mu_tar, log_var_tar)
        # print('recons_loss', recons_loss.item())
        ep_loss += loss.item()
        ep_recon_loss += recons_loss.item()
        ep_reg_loss += re_loss.item()
    # print('\nno.of ietrations in this epoch are ', count, end = '')
    model_optim.zero_grad()
    loss.backward()
    # for name, params in model.named_parameters():
    #     print('gradients', params.grad.mean())
    #     print('weights', params.data)
    model_optim.step()

    tot_loss.append(ep_loss)
    tot_r_loss.append(ep_recon_loss)
    tot_mse_loss.append(ep_reg_loss)
    print("Epoch:", '%04d' % (ep + 1), "train_loss=", "{:.5f}".format(ep_loss),\
            "recon_loss=", "{:.5f}".format(ep_recon_loss),"time=", "{:.5f}".format(time.time() - t))
plt.title('tot_loss')
plt.plot(tot_loss)
plt.show()

plt.title('recons_loss')
plt.plot(tot_r_loss)
plt.show()

plt.title('kld loss')
plt.plot(tot_mse_loss)
plt.show()

```

**Reply by rusty1s** [reply]
Encode would still operate on the same node feature matrix, but will use different GNN layers for learning two representations which you can then lift to source and destination node representations.

**Reply by akpas001** [reply]
Thank you! That worked! But,as far as I can think we will not be able to apply the concept of variational graph autoencoders to this. Because, it is creating the problem of calculating KL Divergence loss for two latent vectors z1 and z2. Is there any way around it?

**Reply by rusty1s** [reply]
Why isn't that possible? The KL divergence is always only computed on individual vectors. Let me know if I am missing something.
